scripts/extensions/sams/package.json

```diff
@@ -0,0 +1,26 @@
+{
```

Comment:
Could you drop redux/reselect? Redux might be helpful for managing the global app state, but this is a single CRUD view. Adding redux/reselect only creates a layer of indirection and a pile of boilerplate to manage without any benefits I can see.

You could use a simple react component for separating data management:

```
interface ISet {
    _id: string;
    // ...
}

interface IProps {

}

interface IState {
    sets: Array<ISet>;
}

class SetsDataLayerComponent extends React.Component<IProps, IState> {
    constructor(props: IProps) {
        super(props);

        this.state = {
            sets: [],
        };

        this.addSet = this.addSet.bind(this);
        this.removeSet = this.removeSet.bind(this);
    }

    addSet() {
        this.setState({sets: this.state.sets.concat({_id: 'newid'})});
    }

    removeSet(id: string) {
        this.setState({sets: this.state.sets.filter(({_id}) => _id !== id)});
    }

    render() {
        return (
            <SetsManagementUi
                addSet={this.addSet}
                removeSet={this.removeSet}
            />
        );
    }
}
```

Reply:
This is the first of many PRs for SAMS integration into Superdesk. This workspace, for example, is going to have a lot more functionality. Loading Sets/StorageDestination into the store and using selectors can help reduce boilerplate in future sections of this UI, and also separate the state and view. This method was used heavily in Planning and proved very useful. It also helps keeps the view in sync with the data, i.e. update a set in the store and all views that use that set are automatically updated

Reply:
> Loading Sets/StorageDestination into the store and using selectors can help reduce boilerplate in future sections of this UI, and also separate the state and view.

Regarding separation between state and view, the 50 LOC data layer component I posted already does it without requiring 10 files of boilerplate or external libraries. The same data layer component can be used for multiple UIs. Data can be synced by updating the data layer component state on WebSocket events. If there's a need for common code we can extract code to functions/components and reuse them this way.

I didn't work on planning much, but when I did I found it annoying and a waste of time to have to go through all the levels of abstraction only to discover a straight-forward setting of a value which could have been a one or two calls instead of 7.

Do you have an example from planning where by using redux/reselect you have been able to do something which wouldn't be possible using a simple wrapper component?

I'd say let's start with a minimalistic approach using the data component and introduce something more complex only if we hit into obstacles.